<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>DeBasher</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <!-- begin TOC stuff -->
    <link rel="stylesheet" href="stylesheets/toc.css">
    <script src="javascripts/jquery-2.1.3.min.js"></script>
    <script src="javascripts/jquery.toc.min.js"></script>
    <script>
      $(function(){
        $('#toc').toc({'container': 'section'});
      });
    </script>
    <!-- end TOC stuff -->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <!-- <p style="text-align:left;">  <IMG SRC="images/my_logo_panpipe.png" Alt="DeBasher" WIDTH=165> </p> -->
        <p> <font size="3">DeBasher</font></p>
        <div id="toc"></div>
        <p class="view"><a href="https://github.com/daormar/debasher">View the Project on GitHub <small>daormar/debasher</small></a></p>

      </header>
      <section>
        <p style="font-size:30px; font-weight:bold; color:black">DeBasher</p>

<h2>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h2>

<p>DeBasher is a software package implementing a flow-based programming
extension for Bash. In the flow-based programming paradigm, networks of
black box processes with well defined data exchange connections are
executed. DeBasher is characterized by a strong orientation towards the
definition of software components or modules.</p>

<p>DeBasher can be used to execute general programs. However, DeBasher
is particularly appropriate for the execution of pipelines, also called
workflows. Workflow execution is important in different fields. One
particular example would be the field of Bioinformatics.</p>

<p>The DeBasher software package was created and is currently maintained
by <a href="http://daormar.github.io/">Daniel Ortiz-Martínez</a>.
Daniel is an experienced machine learning researcher and also holds
an MSc degree in Bioinformatics.</p>

<!-- <h2> -->
<!-- <a id="news" class="anchor" href="#news" aria-hidden="true"><span class="octicon octicon-link"></span></a>News</h2> -->
<!-- <p>A new version of the toolkit has been released with several improvements and new features: -->
<!-- <ul> -->
<!-- <li>TBD</li> -->
<!-- </ul> -->

<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<p>The package incorporates the following features:</p>

<ul>
<li>Definition of complex programs composed of several processes.</li>
<li>Framework to define programs and processes using Bash shell scripting or other programming languages.</li>
<li>Concurrent execution of processes depending on detailed process dependencies.</li>
<li>Reexecution of only those processes that are pending or whose execution failed.</li>
<li>Integration with workload schedulers such as Slurm.</li>
<li>Support for Conda environments.</li>
<li>Generation of graphical representations of programs as directed acyclic graphs.</li>
</ul>

<h2>
<a id="distribution-details" class="anchor" href="#distribution-details" aria-hidden="true"><span class="octicon octicon-link"></span></a>Distribution Details</h2>

<p>DeBasher has been coded using Bash shell-scripting and Python. DeBasher
is known to compile on Unix-like and Windows (using Cygwin) systems. See
the "Documentation and Support" section of these instructions if you
experience problems during compilation.</p>

<p>It is released under the <a href="http://www.gnu.org/copyleft/lgpl.html">GNU Lesser General Public License (LGPL)</a>.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<h4>
<a id="basic-installation-procedure" class="anchor" href="#basic-installation-procedure" aria-hidden="true"><span class="octicon octicon-link"></span></a>Basic Installation Procedure</h4>

<p>To install DeBasher, first you need to install the autotools
(autoconf, autoconf-archive, automake and libtool packages in
Ubuntu). DeBasher requires Bash 4.0 or above as well as Python 3.x to
work. If you are planning to use DeBasher on a Windows platform, you
also need to install the <a href="https://www.cygwin.com/">Cygwin</a>
environment. Alternatively, the tool can also be installed on Mac OS X
systems using <a href="https://www.macports.org/">MacPorts</a>. Finally,
the Graphviz package is also required so as to generate graphic
information about programs.
</p>

<p> Assuming Ubuntu is being used, the required packages can be
installed as follows:</p>

<pre><code> $ sudo apt install autoconf autoconf-archive automake libtool graphviz
</code></pre>

<p>On the other hand, some of the functionality incorporated by DeBasher
requires the previous installation of third-party software
(<a href="#third-party-software">see below</a>).</p>

<p>Once the autotools are available (as well as other required software
such as Cygwin or MacPorts), you can proceed with the installation of
the tool by following the next sequence of steps:</p>

<ol>
<li>
<p>Obtain the package using git:</p>

<pre><code>$ git clone https://github.com/daormar/debasher.git
</code></pre>

<p>Or <a href="https://github.com/daormar/debasher/archive/master.zip">download it in a zip file</a></p>
</li>
<li><p><code>cd</code> to the directory containing the package's source code and type
<code>./reconf</code>.</p></li>
<li><p>Type <code>./configure</code> to configure the package.</p></li>
<li><p>Type <code>make</code> to compile the package.</p></li>
<li><p>Type <code>make install</code> to install the programs and any data files and
documentation.</p></li>
<li><p>You can remove the program binaries and object files from the source
code directory by typing <code>make clean</code>.</p></li>
</ol>

<p>By default the files are installed under the <code>/usr/local</code>
directory (or similar, depending on the OS you use); however, since Step
5 requires root privileges, another directory can be specified during
Step 3 by typing:</p>

<pre><code> $ configure --prefix=&lt;absolute-installation-path&gt;
</code></pre>

<p>For example, if <code>user1</code> wants to install the DeBasher
package in the directory <code>/home/user1/debasher</code>, the sequence
of commands to execute should be the following:</p>

<pre><code> $ ./reconf
 $ configure --prefix=/home/user1/debasher
 $ make
 $ make install
</code></pre>

<p>The installation directory can be the same directory where the
DeBasher package was decompressed.</p>

<!-- <h4 id="checking-package-installation">Checking Package Installation</h4> -->

<!-- <p>Once the package has been installed, it is possible to perform basic -->
<!-- checkings in an automatic manner so as to detect portability errors. For -->
<!-- this purpose, the following command can be used:</p> -->

<!-- <pre><code> $ make installcheck</code></pre> -->

<!-- <p>TBD</p> -->

<h4>
<a id="third-party-software" class="anchor" href="#third-party-software"
aria-hidden="true"><span class="octicon octicon-link"></span></a>Third
Party Software</h4>

<h5>Slurm</h5>

<p>DeBasher can be configured to
use <a href="https://slurm.schedmd.com/">Slurm</a> as a workload
scheduler. Slurm is particularly indicated to execute large pipelines or
to execute pipelines in high performance computing environments.</p>

<h5>Conda</h5>

<p>DeBasher provides support for automated installation
of <a href="https://conda.io/">Conda</a> packages. Such packages are
organized in environments and (optionally) used within DeBasher software
modules.</p>

<h5>Docker</h5>

<p>DeBasher also provides support for <a href="https://www.docker.com/">Docker</a>
containers.</p>

<h2>
<a id="current-status" class="anchor" href="#current-status" aria-hidden="true"><span class="octicon octicon-link"></span></a>Current Status</h2>

<p> The DeBasher package is under development. Below there is a list of
  features that will be incorporated in a near future:</p>

<ul>
<li>Graphic tool to build programs.</li>
<li>Support for additional workload schedulers.</li>
</ul>

<h2>
<a id="docsupport" class="anchor" href="#docsupport" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample Projects Using DeBasher</h2>
<p></p>
<ul>
<li><p><a href="https://daormar.github.io/geno-debasher/">Geno-DeBasher</a>:
Geno-DeBasher is a bioinformatics software project implementing DeBasher
modules related to genome variant calling.</p></li>
</ul>

<h2>
<a id="docsupport" class="anchor" href="#docsupport" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation and Support</h2>

<p>Project documentation is being developed. Such documentation includes:</p>

<ul>
<li>These instructions.</li>
<li>A set
of <a href="http://daormar.github.io/debasher/docsupport/slides.pdf">slides</a>
("slides.pdf" under the "doc" directory).</li>
<li><a href="https://debasher.readthedocs.io/en/latest/">Technical documentation</a>.</li>
</ul>

<p>If you need additional help, you can:</p>

<ul>
<li>use the <a href="https://github.com/daormar/debasher/issues">github issue tracker</a>.
</li>
<li><a href="mailto:daniel.ortiz.phd@gmail.com">send an e-mail to the author</a>.</li>
</li>
</ul>

<!-- <p>Additional information about the theoretical foundations of Thot can be -->
<!-- found in [Ortiz-Martínez, 2011]. One interesting feature of Thot, -->
<!-- incremental (or online) estimation of statistical models, is also -->
<!-- described in [Ortiz-Martínez et al., 2010]. Finally, phrase-level -->
<!-- alignment generation functionality implemented in Thot was proposed in -->
<!-- [Ortiz-Martínez et al., 2008].</p> -->

<!-- <h2> -->
<!-- <a id="citation" class="anchor" href="#citation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citation</h2> -->

<!-- <p>You are welcome to use the code under the terms of the license for -->
<!-- research or commercial purposes, however please acknowledge its use with -->
<!-- a citation:</p> -->

<!-- <p>Daniel Ortiz-Martínez, Francisco Casacuberta. -->
<!-- <em>"The New Thot Toolkit for Fully Automatic and Interactive Statistical Machine Translation"</em>. -->
<!-- In Proc. of the European Association for Computational Linguistics (EACL): System Demonstrations, -->
<!-- Gothenburg, Sweden, April 2014. pp. 45-48.</p> -->

<!-- <p>Here is a BiBTeX entry:</p> -->

<!-- <pre> -->
<!-- @InProceedings{Ortiz2014, -->
<!--   author    = {Daniel Ortiz-Mart\'{\i}nez and Francisco Casacuberta}, -->
<!--   title     = {The New Thot Toolkit for Fully Automatic and Interactive Statistical Machine Translation}, -->
<!--   booktitle = {Proc. of the European Association for Computational Linguistics (EACL): System Demonstrations}, -->
<!--   year      = {2014}, -->
<!--   month     = {April}, -->
<!--   address   = {Gothenburg, Sweden}, -->
<!--   pages     = "45--48", -->
<!-- } -->
<!-- </pre> -->

<!-- <h2> -->
<!-- <a id="literature" class="anchor" href="#literature" aria-hidden="true"><span class="octicon octicon-link"></span></a>Literature</h2> -->

<!-- <p> Shlomi, T., M. N. Cabili, M. J. Herrgård, B. Ø. Palsson, and Eytan -->
<!-- Ruppin (2008). “Network-based prediction of human tissue-specific -->
<!-- metabolism”. In: Nat. Biotechnol. 26(9), pp. 1003–10.</p> -->

<!-- <p>Gudmundsson, S. and I. Thiele (2010). “Computationally efficient flux -->
<!-- variability analy- sis”. In: BMC Bioinformatics 11, p. 489.</p> -->

<!-- <p> Hebenstreit, D., M. Fang, M. Gu, V. Charoensawan, A. van -->
<!-- Oudenaarden, and S. A. Teichmann (2011). “RNA sequencing reveals two -->
<!-- major classes of gene expression levels in metazoan cells”. In: -->
<!-- Mol. Syst. Biol. 7, p. 497.</p> -->

<!-- <p> Erdrich, P., R. Steuer, and S. Klamt (2015). “An algorithm for the -->
<!-- reduction of genome-scale metabolic network models to -->
<!-- meaningful core models”. In: BMC Syst Biol 9, p. 48.</p> -->

<!-- <h2> -->
<!-- <a id="sponsors" class="anchor" href="#sponsors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sponsors</h2> -->

<!-- <p>Thot has been supported by the European Union under the <a href="http://www.casmacat.eu/">CasMaCat -->
<!-- research project</a>. Thot has also received support from the Spanish -->
<!-- Government in a number of research projects, such as the <a href="http://miprcv.iti.upv.es/">MIPRCV -->
<!-- project</a> that belonged to the prestigious CONSOLIDER programme</a>.</p> -->

<!--  <ul class="partner-logos"> -->
<!--     <li><a href="http://cordis.europa.eu/fp7/home_en.html"><img src="http://daormar.github.io/thot/images/jaune.jpg" alt="EU" height="50" /></a></li> -->
<!--     <li><a href="http://www.micinn.es/"><img src="http://daormar.github.io/thot/images/logo-mci.jpg" alt="MCINN" /></a></li> -->
<!--     <li><a href="http://www.vlc-campus.com"><img height="50" src="http://daormar.github.io/thot/images/vlc-campus.png" alt="VLC-CAMPUS" /></a></li> -->
<!--     <li><a href="http://www.upv.es"><img src="http://daormar.github.io/thot/images/logo-upv-hor.png" alt="UPV" height="50" /></a></li> -->
<!--   </ul> -->

<!-- <p>Last updated: 25 July 2017</p> -->
<!--       </section> -->
<!--     </div> -->
<!--     <script src="javascripts/scale.fix.js"></script> -->
<!--               <script type="text/javascript"> -->
<!--             var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www."); -->
<!--             document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E")); -->
<!--           </script> -->
<!--           <script type="text/javascript"> -->
<!--             try { -->
<!--               var pageTracker = _gat._getTracker("UA-60880705-1"); -->
<!--             pageTracker._trackPageview(); -->
<!--             } catch(err) {} -->
<!--           </script> -->

  </body>
</html>
